src/renderer/gdi/paint.cpp

```diff
@@ -284,13 +284,13 @@ using namespace Microsoft::Console::Render;
 // See: Win7: 390673, 447839 and then superseded by http://osgvsowi/638274 when FE/non-FE rendering condensed.
 //#define CONSOLE_EXTTEXTOUT_FLAGS ETO_OPAQUE | ETO_CLIPPED
 //#define MAX_POLY_LINES 80
-[[nodiscard]] HRESULT GdiEngine::PaintBufferLine(std::basic_string_view<Cluster> const clusters,
+[[nodiscard]] HRESULT GdiEngine::PaintBufferLine(const RenderClusterIterator clusterIter,
                                                  const COORD coord,
                                                  const bool trimLeft) noexcept
 {
     try
     {
-        const auto cchLine = clusters.size();
+        size_t cchLine = 120;

```

Comment:
Again, without size I can't make use of length to init arrays.

Reply:
Again, decent estimate is coord to right edge of dirty rect.

Reply:
An estimate is not enough, though. I need the *exact* size here.

Reply:
Iterating it two times, could it be that bad? I'm stucked...

Reply:
Not strictly. You could replace things like `auto pwsPoly = std::make_unique<wchar_t>(cchLine)` instead with a `std::wstring` and `.reserve()` it up to the "estimate" of `_rcInvalid.Right - coord.X`. Same goes for the `rgdxPoly` with a vector of ints or something. (or use vectors for both).

The first phase with the `for` loop has to iterate the whole thing anyway to condense the clusters back into a string. Either you'll get lucky in the majority case and the estimate won't resize the vectors, or you'll get unlucky in the uncommon case and you'll resize up once or twice. (And if resizing up happens a lot, we can use an estimate function like the `CustomTextLayout::_EstimateGlyphCount` does to reduce the potential resizes.)

By the end of the `for` loop (lines 316-325), you should be able to know the exact widths required for all the functions further down `PaintBufferLine` and have only iterated once. E.g. you should have calculated `cchLine` exactly by iterating once and also filled a vector `pwsPoly` and vector `rgdxPoly` with the condensed version. That should be enough information to finish the draw operation.

However, now that I look at this further. The loop in this specific GDI renderer only allows a maximum of one `wchar_t` per unit because of GDI limitation. (comment on L322 where `cluster.GetTextAsSingle()` is called. So the "estimate" used for `.reserve()` above should actually be the exact column count as a `Cluster` is a console grid unit described by one-or-more-`wchar_t` and the vectors wouldn't auto-resize.

The last problem is that the vectors would free when going out of scope. There isn't really a `detach` to the vectors. We could just hold a reference to them that is freed inside `_FlushBufferLines` after `PolyTextOut` is called. 

Member variables on the class like:
`std::vector<std::vector<wchar_t>> _polyStrings` and `std::vector<std::vector<int>> _polyWidths`, push a new vector into those on each `PaintBufferLine`, take the `.data()` of the line and stuff it into the `pPolyTextLine` fields, and then clear them out on `_FlushBufferLines`.

That might make `_FlushBuferLines` a little cleaner than raw `delete[]`s of bare pointers as well.